[Backtracing] Add support for looking up paths for auxiliary executables

al45tair opened the pull request:
We need to be able to locate `swift-backtrace` relative to the current location of the runtime library.

This needs to work:

* In a Swift build directory.
* On Darwin, where we're installed in `/usr/lib/swift` and `/usr/libexec/swift`.
* On Linux, where we're in `/usr/lib/swift/linux` and `/usr/libexec/swift/linux`.
* On Windows, where we may be in a flat directory layout (because of limitations of Windows DLL lookups).

rdar://103071801

al45tair committed:
[Backtracing] Add support for looking up paths for auxiliary executables.

We need to be able to locate `swift-backtrace` relative to the current
location of the runtime library.

This needs to work:

* In a Swift build directory.
* On Darwin, where we're installed in /usr/lib/swift and /usr/libexec/swift.
* On Linux, where we're in /usr/lib/swift/linux and /usr/libexec/swift/linux.
* On Windows, where we may be in a flat directory layout (because of limitations
  of Windows DLL lookups).

rdar://103071801

al45tair commented:
This is part of a set of stacked PRs:

|  | PR                | Title                                                                      | Merges Into       |
|--|:------------------|:---------------------------------------------------------------------------|:------------------|
|  | apple/swift#63650 | [Backtracing] Add support for building target executables into libexec     | main              |
|👉| al45tair/swift#8  | [Backtracing] Add support for looking up paths for auxiliary executables   | apple/swift#63650 |
|  | al45tair/swift#9  | [Backtracing][Runtime][Darwin] Add code to trigger the external backtracer | al45tair/swift#8  |
|  | al45tair/swift#10 | [Backtracing] Add _Backtracing library to the build                        | al45tair/swift#9  |
|  | al45tair/swift#11 | [Backtracing][Frontend] Add support for implicit import of _Backtracing    | al45tair/swift#10 |
|  | al45tair/swift#12 | [Backtracing] Add swift-backtrace to the build                             | al45tair/swift#11 |
|  | al45tair/swift#13 | [Backtracing][Docs] Add an explanation of the workings of the backtracer   | al45tair/swift#12 |
|  | al45tair/swift#14 | [Backtracing] Add control over symbol caching                              | al45tair/swift#13 |

(Only the bottom PR will be able to run PR checks at any time. Please don't wait for the checks before review.)

al45tair commented:
The original version of this is at apple/swift#62462 and various folks have looked at this before.

grynspan commented on include/swift/Runtime/EnvironmentVariables.h:
```diff
@@ -24,6 +24,10 @@ void initialize(void *);
 
+// Define a typedef "string" in swift::runtime::environment to make string
+// environment variables work
+using string = const char *;
```
wat

al45tair replied:
Yeah, I know. It's because someone went mad with macros with the environment variable stuff. I'd like to fix that to use fewer macros and a bit more C++, which would remove the need for this horrifying line of code, but that's too big a thing to do as part of another piece of work.

grynspan commented on include/swift/Runtime/Paths.h:
```diff
@@ -0,0 +1,77 @@
+/// \return A string containing the full path to the executable.
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getAuxiliaryExecutablePath(const char *name);
```
Presumably this string needs to be cleaned up somehow, right?

al45tair replied:
Yes, it'll need to be `std::free()`d when it isn't needed any more. (I'll update the comment.)

grynspan replied:
This one should be `copy` in this case (and return `char *`, not `const char *`, so the result can be passed to `free()`.)

grynspan commented on stdlib/public/runtime/EnvironmentVariables.cpp:
```diff
@@ -24,6 +25,12 @@ using namespace swift;
 
+// This is required to make the macro machinery work correctly; we can't
+// declare a VARIABLE(..., const char *, ...) because then the token-pasted
+// names won't work properly.  It *does* mean that if you want to use std::string
```
Maybe name it something unlikely to collide, like `swift_const_char_ptr`?

al45tair replied:
That would be fine, except that then the documentation you get from `SWIFT_HELP` will tell you the type is `const_char_ptr`. Like I said, someone went a bit OTT with macros.

grynspan commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,554 @@
+}
+
+bool
+_swift_lookingAtBin(const char *ptr, const char *base)
```
These helper functions are internal to this file, right? If so, they can be made `static` and don't need the `_swift_` prefix.

al45tair replied:
It's inside an anonymous namespace already, so no need for `static`. It probably doesn't need the `_swift_` prefix, but that part won't hurt.

grynspan commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,554 @@
+      pszDrive[len - 1] = 0;
+
+    WCHAR ntPath[4096];
+    dwRet = QueryDosDeviceW(pszDrive, ntPath, 4096);
```
_`TARGET_OS_DOS` intensifies_

al45tair committed:
[Backtracing] Update the comment for swift_getAuxiliaryExecutablePath()

This should have said how to release the string.

rdar://103071801

mikeash commented on include/swift/Runtime/Paths.h:
```diff
@@ -0,0 +1,78 @@
+///         owned by the runtime and should not be freed.
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRuntimePath();
```
`swift_getRuntimeLibraryPath` might be less confusing in the long-term. The use of "runtime" to describe the library isn't entirely obvious otherwise.

al45tair replied:
That's not a bad idea. I'll update that tomorrow.

mikeash commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,554 @@
+  //                         +-------------- ptr - 10
+
+  return (ptr - base >= 10
+          && _swift_isPathSep(ptr[-10])
```
I think this is the first time I've ever seen a negative array index in C that wasn't deliberately breaking something.

al45tair replied:
:smile:

mikeash commented on stdlib/public/runtime/Paths.cpp:
```diff
@@ -0,0 +1,554 @@
+  path = dli.dli_fname;
+#endif
+
+  runtimePath = ::strdup(path);
```
Do we need to do any canonicalization here? I think that probably gets done for us by dyld but not sure about other platforms.

al45tair replied:
I don't think it's an issue in practice. At worst, we won't be able to locate `swift-backtrace` if something funny is going on, but that isn't a huge problem.

AFAIK `dladdr()` always returns a full path (it's documented as returning a "pathname"). It might have symlinks in it, but I don't think we care about that — we just want to try to resolve relative to the symlinked path in that case.

al45tair replied:
(Saleem and I had already had a discussion about the situation on Windows, which predictably is the awkward case.)

grynspan commented on include/swift/Runtime/Paths.h:
```diff
@@ -0,0 +1,78 @@
+///         owned by the runtime and should not be freed.
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRuntimePath();
```
I thought I'd commented already, but… these strings are `strdup()`ed in the implementation, so the functions should probably be named `swift_copyX()` instead of `swift_getX()`, and the caller should be expected to free the results. Right?

grynspan replied:
Never mind, I see in the implementation you `once` 'em. If any _do_ produce unique strings per-call, the above applies though.

al45tair replied:
Quite right. That _might_ apply to the function that resolves the name for an auxiliary executable, if I managed to call it `get` instead of `copy` by mistake (which I may have; I will check and fix it if that's what I did).

al45tair committed:
[Backtracing] Rename a couple of functions, some comment changes.

Renamed a couple of functions following review feedback.  Also added some
extra words to the comments.

rdar://103071801

al45tair commented:
Duplicated to apple/swift#63964 to PR test and merge.

al45tair committed:
[Backtracing] Add missing void cast to the paths test.

We need to cast in order to call `free()`.

rdar://103071801

al45tair committed:
[Backtracing] Add target-arch to lit.cfg.

This must have ended up in the wrong PR in the stack.  It is needed in
this one so that we can locate libswiftCore.

rdar://103071801

al45tair committed:
[Backtracing] Change the return type of one of the paths functions.

Since it's copying, it may as well return `char *` instead of
`const char *`.

rdar://103071801

al45tair committed:
[Backtracing] Fix some const problems as a result of the previous change.

rdar://103071801

al45tair committed:
[Backtracing] A couple more const fixes.

rdar://103071801
